### ISSUE TITLE
iter_batched bug?

### ISSUE THREAD

**Issue by veniamin-ilmer** [bheisler/criterion.rs#300]
Please let me know if I am understanding this function incorrectly..

As far as I understood, the `iter_batched` function is used to be able to run something before each iteration that shouldn't be timed.

I started running into inconsistent answers when I tried using this function, so I decided to test it out.

    use criterion::*;

    criterion_group!(test, timing);
    criterion_main!(test);

    fn timing(c: &mut Criterion) {
      use std::{thread, time};

      let ten_millis = time::Duration::from_millis(10);
  
      c.bench_function("timing", move |b| {
        b.iter_batched(|| {
          thread::sleep(ten_millis);
          1
        }, |data| black_box(data) + 1,
        BatchSize::SmallInput)
      });
    }

Gave me this solution:

        Finished release [optimized] target(s) in 1.18s
         Running target\release\deps\test-442db35225cecf43.exe
    timing                  time:   [296.79 ns 332.20 ns 377.48 ns]
    Found 2 outliers among 100 measurements (2.00%)
      1 (1.00%) high mild
      1 (1.00%) high severe

If I remove the `thread::sleep`, I get this:

        Finished release [optimized] target(s) in 1.17s
         Running target\release\deps\test-442db35225cecf43.exe
    Gnuplot not found, disabling plotting
    timing                  time:   [1.8746 ns 1.8918 ns 1.9167 ns]
                            change: [-99.768% -99.721% -99.665%] (p = 0.00 < 0.05)
                            Performance has improved.
    Found 11 outliers among 100 measurements (11.00%)
      7 (7.00%) low severe
      3 (3.00%) high mild
      1 (1.00%) high severe

I don't understand, I thought `iter_batched` is supposed to not include the initial pre-iteration steps into the timing. Am I using the function incorrectly?

Thank you!

**Comment by veniamin-ilmer** [comment]
I see there there are multiple functions that do a similar functionality, I tried out `iter_with_large_setup`:

    use criterion::*;

    criterion_group!(test, timing);
    criterion_main!(test);

    fn timing(c: &mut Criterion) {
      use std::{thread, time};

      let ten_millis = time::Duration::from_millis(10);
      
      c.bench_function("timing", move |b| {
        b.iter_with_large_setup(|| {
          thread::sleep(ten_millis);
          1
        }, |data| black_box(data) + 1)
      });
    }

Resulting:

    timing                  time:   [50.011 ns 59.091 ns 71.077 ns]
                            change: [-26.500% +23.033% +98.890%] (p = 0.44 > 0.05)

Yet if I remove `thread::sleep`:

    timing                  time:   [2.0708 ns 2.0849 ns 2.0992 ns]
                            change: [-99.112% -98.750% -98.150%] (p = 0.00 < 0.05)
                            Performance has improved.

**Comment by veniamin-ilmer** [comment]
From further testing, I see that the amount of timing is similar regardless of how much time I sleep for.

For some reason, the timing is slower simply from use of extra functions, not the actual sleeping of itself.

**Comment by bheisler** [comment]
Hey, thanks for the bug report.

I think this is one of those cases where I just have to say "welcome to the wild world of microbenchmarks". You're right that whether you sleep or not in the setup function should, in principle, have no effect on the measurements. Those sleeps should all have been executed before the actual timing loop starts.

In practice, though, optimizing compilers tend to be extremely sensitive to apparently-meaningless changes in the code. That's why the numbers Criterion.rs (or any benchmarking tool, really) provides can only ever be an estimate. It's an estimate because it's based on statistical calculations, yes, but it's also an estimate because we can only ever measure the function after it's been compiled with whatever set of optimization decisions the compiler made. This may not be the same set of optimization decisions the compiler makes when compiling your code for production. We programmers only have limited tools (such as `black_box`) to control those optimization decisions.

In short - it's likely that including or removing the sleep is causing the compiler to optimize the measurement loop differently, and (as far as I can tell) there isn't anything Criterion.rs can do about that.
